### PR TITLE
Secondary Navigation Menu should scroll when item is not visible

### LINKED ISSUE #250: Prevent secondary navigation jumping when selecting an item
# Bug

## Description
When a secondary navigation item already visible on screen is selected, it should not jump to the top of secondary navigation.

## Additional Context / Screenshots
dev-site-nav-jump

### PR DESCRIPTION
**Pull request by pranav300**
### Summary

Updated secondary navigation menu `scrollIntoView` condition to work when the item is not visible and on top.

Closes #250

### Deployment Link

https://terra-dev-site-deployed-pr-#.herokuapp.com/

### Additional Details

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: src/navigation/_CollapsingNavigationMenu.jsx [hunk 1]
```diff
@@ -120,9 +120,10 @@ class CollapsingNavigationMenu extends React.Component {
 
   componentDidUpdate() {
     const { isNewSelectedPath } = this.state;
-    const currentItemPosition = this.selectedItem && this.selectedItem.current ? this.selectedItem.current.getBoundingClientRect().bottom : null;
-    const navigationMenuHeight = document.querySelector('#terra-dev-site-nav-menu').getBoundingClientRect().bottom;
-    if (isNewSelectedPath && currentItemPosition > navigationMenuHeight) {
+    const currentItemPosition = this.selectedItem && this.selectedItem.current ? this.selectedItem.current.getBoundingClientRect() : null;
+    const navigationMenuPosition = document.querySelector('#terra-dev-site-nav-menu').getBoundingClientRect();
+    if (isNewSelectedPath && currentItemPosition && navigationMenuPosition && (currentItemPosition.bottom > navigationMenuPosition.bottom
```
Comment (ShettyAkarsh) [inline src/navigation/_CollapsingNavigationMenu.jsx:125]: can this condition be like this 
`if (isNewSelectedPath && (currentItemPosition.bottom > navigationMenuPosition.bottom
      || currentItemPosition.top < navigationMenuPosition.top))`  ? do we need  currentItemPosition and navigationMenuPosition here in this condition ?

Comment (pranav300) [reply src/navigation/_CollapsingNavigationMenu.jsx:125]: We need these, as the code may break if any of them is null.

File: src/navigation/_CollapsingNavigationMenu.jsx [hunk 2]
```diff
@@ -120,9 +120,10 @@ class CollapsingNavigationMenu extends React.Component {
 
   componentDidUpdate() {
     const { isNewSelectedPath } = this.state;
-    const currentItemPosition = this.selectedItem && this.selectedItem.current ? this.selectedItem.current.getBoundingClientRect().bottom : null;
-    const navigationMenuHeight = document.querySelector('#terra-dev-site-nav-menu').getBoundingClientRect().bottom;
-    if (isNewSelectedPath && currentItemPosition > navigationMenuHeight) {
+    const currentItemPosition = this.selectedItem && this.selectedItem.current ? this.selectedItem.current.getBoundingClientRect() : null;
```
Comment (mjhenkes) [inline src/navigation/_CollapsingNavigationMenu.jsx:123]: you can use optional chaining (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)
```suggestion
    const currentItemPosition = this.selectedItem?.current ? this.selectedItem.current.getBoundingClientRect() : null;
```

File: src/navigation/_CollapsingNavigationMenu.jsx [hunk 3]
```diff
@@ -120,9 +120,10 @@ class CollapsingNavigationMenu extends React.Component {
...
+    const navigationMenuPosition = document.querySelector('#terra-dev-site-nav-menu').getBoundingClientRect();
+    if (isNewSelectedPath && currentItemPosition && navigationMenuPosition && (currentItemPosition.bottom > navigationMenuPosition.bottom
+      || currentItemPosition.top < navigationMenuPosition.top)) {
       this.selectedItem.current.scrollIntoView();
     }
   }
```
Comment (mjhenkes) [inline src/navigation/_CollapsingNavigationMenu.jsx:129]: I suggest something like this for readability.

```jsx
  componentDidUpdate() {
    const { isNewSelectedPath } = this.state;
    if(!isNewSelectedPath){
      return;
    }
    const currentItemPosition = this.selectedItem?.current ? this.selectedItem.current.getBoundingClientRect() : null;
    const navigationMenuPosition = document.querySelector('#terra-dev-site-nav-menu').getBoundingClientRect();
    if(!isNewSelectedPath || !navigationMenuPosition){
      return;
    }
    // If the current item is not visible, scroll the item into view.
    if ((currentItemPosition.bottom > navigationMenuPosition.bottom
      || currentItemPosition.top < navigationMenuPosition.top)) {
      this.selectedItem.current.scrollIntoView();
    }
  }
```

File: tests/wdio/secondaryNavigation-spec.js [hunk 4]
```diff
@@ -18,3 +18,41 @@ Terra.describeViewports('secondary scroll multiple directories', ['huge'], () =>
...
+    before(() => browser.url('/secondary-nav-test/terra-dev-site/secondary-scroll-test/filler/filler-test'));
+    Terra.it.validatesElement('Before', { selector: '#root' });
+    it('selects filler 04 Test', () => {
+      browser.click('[class*="Extension-module__extension"]');
+      browser.keys('Filler 04');
+      browser.waitForVisible('[class*="list-item"]');
+      browser.click('[class*="list-item"]');
+    });
+    Terra.it.validatesElement('After', { selector: '#root' });
+  });
```
Comment (mjhenkes) [inline tests/wdio/secondaryNavigation-spec.js:33]: I think it's cleaner to contain your test to a single it block.
```suggestion
  it('should not scroll when item is visible', () => {
    browser.url('/secondary-nav-test/terra-dev-site/secondary-scroll-test/filler/filler-test'));
    Terra.validatesElement('Before', { selector: '#root' });
    browser.click('[class*="Extension-module__extension"]');
    browser.keys('Filler 04');
    browser.waitForVisible('[class*="list-item"]');
    browser.click('[class*="list-item"]');
    Terra.validatesElement('After', { selector: '#root' });
  });
```

Comment (pranav300) [reply tests/wdio/secondaryNavigation-spec.js]: Updated here https://github.com/cerner/terra-dev-site/pull/261/commits/f7cc5a3c33dc0374e91fefb452b76a895a8e7543.

File: src/navigation/_CollapsingNavigationMenu.jsx (hunk 3 above)
Comment (pranav300) [reply src/navigation/_CollapsingNavigationMenu.jsx]: Updated here https://github.com/cerner/terra-dev-site/pull/261/commits/f7cc5a3c33dc0374e91fefb452b76a895a8e7543

File: src/navigation/_CollapsingNavigationMenu.jsx (hunk 2 above)
Comment (pranav300) [reply src/navigation/_CollapsingNavigationMenu.jsx]: Updated here https://github.com/cerner/terra-dev-site/pull/261/commits/f7cc5a3c33dc0374e91fefb452b76a895a8e7543

File: src/navigation/_CollapsingNavigationMenu.jsx [hunk 5]
```diff
@@ -120,9 +120,16 @@ class CollapsingNavigationMenu extends React.Component {
 
   componentDidUpdate() {
     const { isNewSelectedPath } = this.state;
-    const currentItemPosition = this.selectedItem && this.selectedItem.current ? this.selectedItem.current.getBoundingClientRect().bottom : null;
-    const navigationMenuHeight = document.querySelector('#terra-dev-site-nav-menu').getBoundingClientRect().bottom;
-    if (isNewSelectedPath && currentItemPosition > navigationMenuHeight) {
+    if (!isNewSelectedPath) {
+      return;
+    }
+    const currentItemPosition = this.selectedItem?.current ? this.selectedItem.current.getBoundingClientRect() : null;
```
Comment (supreethmr) [inline src/navigation/_CollapsingNavigationMenu.jsx:126]: curious to know what role does `?` plays here `this.selectedItem?`.

Comment (yuderekyu) [reply src/navigation/_CollapsingNavigationMenu.jsx:126]: optional chaining https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

Comment (emilyrohrbough) [reply src/navigation/_CollapsingNavigationMenu.jsx:126]: 🤯 that's so cool! Learned something new.

**Comment by emilyrohrbough** [comment]
@pranav300  nice job!

**Pull request merged**